src_lockfree/size.ml

```diff
@@ -0,0 +1,219 @@
+(* Copyright (c) 2023 Vesa Karvonen
+
+   Permission to use, copy, modify, and/or distribute this software for any
+   purpose with or without fee is hereby granted, provided that the above
+   copyright notice and this permission notice appear in all copies.
+
+   THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
+   REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
+   AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
+   INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
+   LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
+   OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
+   PERFORMANCE OF THIS SOFTWARE. *)
+
+(** ⚠️ Beware that this implementation uses a bunch of low level data
+    representation tricks to minimize overheads. *)
+
+module Atomic = Transparent_atomic
+
+let max_value = Int.max_int
+
+module Snapshot = struct
+  type t = int Atomic.t array
+  (** We use an optimized flat representation where the first element of the
+      array holds the status of the snapshot.
+
+        +--------+---------+---------+---------+- - -
+        | status | counter | counter | counter | ...
+        +--------+---------+---------+---------+- - -
+
+      The status is either {!collecting}, {!computing}, or a non-negative value.
+
+      The counter snapshot values are initialized to a negative value and after
+      collecting they will all be non-negative. *)
+
+  let zero = [| Atomic.make 0 |]
+  let collecting = -1
+  let computing = -2
+
+  let[@inline] is_collecting (s : t) =
+    Atomic.get (Array.unsafe_get s 0) = collecting
+
+  let create n = Array.init n @@ fun _ -> Atomic.make collecting
+
+  let[@inline] set s i after =
+    let snap = Array.unsafe_get s i in
+    let before = Atomic.get snap in
+    if
+      before = collecting
+      || (* NOTE: The condition below accounts for overflow. *)
+      (after - before - 1) land max_value < max_value / 2
```

Comment:
The overall goal at this point (and in the similar snippet of code in `forward`) is to update the copy of a counter in the snapshot with the _latest_ value of the counter.

The counters are always only incremented.  So, typically we want to update when `after > before`. Unfortunately it is possible for a counter to overflow:
```ocaml
# Int.max_int + 1 ;;
- : int = -4611686018427387904
```
The condition above is written in such a way that the copy is (practically always) updated correctly even in case of overflow (unless the counter is incremented roughy `max_value / 2` times, which is extremely unlikely).  Here is a REPL snippet that shows results from the condition:
```ocaml
# let max_value = Int.max_int ;;
val max_value : int = 4611686018427387903
# let condition before after = (after - before - 1) land max_value < max_value / 2 ;;
val condition : int -> int -> bool = <fun>
# let plus n delta = (n + delta) land max_value ;;
val plus : int -> int -> int = <fun>
# let try_cond before delta = condition before (plus before delta) ;;
val try_cond : int -> int -> bool = <fun>
# try_cond max_value 0 ;;
- : bool = false
# try_cond max_value 1 ;;
- : bool = true
# try_cond max_value 2 ;;
- : bool = true
# try_cond max_value 3 ;;
- : bool = true
# try_cond max_value (-1) ;;
- : bool = false
# try_cond max_value (-2) ;;
- : bool = false
# try_cond max_value (-3) ;;
- : bool = false
# try_cond max_value (max_value / 2) ;;
- : bool = true
# try_cond max_value (max_value / 2 + 1) ;;
- : bool = false
# try_cond max_value (max_value / 2 - 1) ;;
- : bool = true
- : bool = true
# condition 0 (-1) ;;
- : bool = false
# condition Int.max_int (-1) ;;
- : bool = false
# condition Int.max_int (-2) ;;
- : bool = false
# condition 0 (-2) ;;
- : bool = false
```
So, basically, the update is attempted (condition is true) when the new (`after`) value seems more likely to be a new value (than an old value) and is not one the two special negative values.